glib/src/gstring.rs

```diff
@@ -404,6 +404,74 @@ impl<'a> ToGlibPtr<'a, *mut c_char> for GString {
     }
 }
 
+#[doc(hidden)]
+impl<'a> FromGlibContainer<*const c_char, *const i8> for GString {
+    unsafe fn from_glib_none_num(ptr: *const i8, num: usize) -> Self {
+        if num == 0 {
+            return Self::from("");
+        }
+        let mut bytes = Vec::with_capacity(num + 1);
+        let slice = std::slice::from_raw_parts(ptr as *const u8, num);
+        bytes.extend_from_slice(slice);
+        bytes.push(0);
+
+        CStr::from_bytes_with_nul_unchecked(bytes.as_slice()).into()
+    }
+
+    unsafe fn from_glib_container_num(ptr: *const i8, num: usize) -> Self {
+        GString(Inner::Foreign(ptr as *mut _, num))
+    }
+
+    unsafe fn from_glib_full_num(ptr: *const i8, num: usize) -> Self {
+        GString(Inner::Foreign(ptr as *mut _, num))
```

Comment:
I wonder if these work correctly for empty strings or some fixes would be needed. Can you add a test for empty strings in the different variants?

Reply:
I.e. `NULL` / `0` and `$random_pointer` / `0` should give the same results for `none` and `full`

Reply:
I have added more cases to the test